internal/cli/cmd/auth/authorize/authenticator/pkce_authenticator.go

```diff
@@ -0,0 +1,320 @@
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+
+	"github.com/benbjohnson/clock"
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/pkg/errors"
+
+	"github.com/apecloud/kubeblocks/internal/cli/cmd/auth/utils"
+)
+
+type OIDCWellKnownEndpoints struct {
+	AuthorizationEndpoint string `json:"authorization_endpoint"`
+	TokenEndpoint         string `json:"token_endpoint"`
+}
+
+type PKCEAuthenticator struct {
+	client             *http.Client
+	Clock              clock.Clock
+	ClientID           string
+	AuthURL            string
+	AuthAudience       string
+	Challenge          Challenge
+	WellKnownEndpoints *OIDCWellKnownEndpoints
+}
+
+// Challenge holds challenge and verification data needed for the PKCE flow
+type Challenge struct {
+	Code     string
+	Verifier string
+	Method   string
+}
+
+// CallbackResponse holds the code gotten from the authorization callback.
+// Error will hold an error struct if an error occurred.
+type CallbackResponse struct {
+	Code  string
+	Error error
+}
+
+type AuthorizationResponse struct {
+	CallbackURL string
+	Code        string
+}
+
+func newPKCEAuthenticator(client *http.Client, clientID string, authURL string) (*PKCEAuthenticator, error) {
+	if client == nil {
+		client = cleanhttp.DefaultClient()
+	}
+	p := &PKCEAuthenticator{
+		client:       client,
+		Clock:        clock.New(),
+		ClientID:     clientID,
+		AuthURL:      authURL,
+		AuthAudience: authURL + "/api/v2/",
+	}
+	var err error
+	p.WellKnownEndpoints, err = getOIDCWellKnownEndpoints(authURL)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get OIDC well known endpoints")
+	}
+	p.Challenge = defaultChallengeGenerator()
+	return p, nil
+}
+
+func (p *PKCEAuthenticator) GetAuthorization(ctx context.Context, openURLFunc func(URL string), states ...string) (interface{}, error) {
+	callback := newCallbackService("8000")
+	codeReceiverCh := make(chan CallbackResponse)
+	defer close(codeReceiverCh)
+
+	var state string
+	if states == nil {
+		state = defaultStateGenerator()
+	} else {
+		state = states[0]
+	}
+	go callback.awaitResponse(codeReceiverCh, state)
+
+	params := url.Values{
+		"audience":              []string{p.AuthAudience},
+		"client_id":             []string{p.ClientID},
+		"code_challenge":        []string{p.Challenge.Code},
+		"code_challenge_method": []string{p.Challenge.Method},
+		"response_type":         []string{"code"},
+		"state":                 []string{state},
+		"redirect_uri":          []string{callback.getCallbackURL()},
+		"scope": []string{strings.Join([]string{
+			"read_databases", "write_databases", "read_user", "read_organization", "offline_access", "openid", "profile", "email",
+		}, " ")},
+	}
+
+	URL := fmt.Sprintf("%s?%s",
+		p.WellKnownEndpoints.AuthorizationEndpoint,
+		params.Encode(),
+	)
+	openURLFunc(URL)
+
+	callbackResult := <-codeReceiverCh
+
+	if callbackResult.Error != nil {
```

Comment:
Is there a callback leak when error happens ?

Reply:
Sorry, I was on leave last week, so coming back to this question late. According to the reminder, I checked the callback function and the call of the callback, and there is no callback leakage problem. We listened to port 8000 locally and set the callback function /callback. This callback function will be called by Auth0 and pass the Authorization Code. Here we need to check whether the code is passed by Auth0. If no code is passed, the main program will be notified through the channel. An error is detected, and the terminal prints to inform the user that the requested login failed.